Add control structure including a method to factory reset the KNoT Gateway

alla3 opened the pull request:
Add control structure, including a D-Bus method allowing to factory reset the KNoT Gateway.

alla3 committed:
Add initial main structure

Add initial main stucture, including glib mainloop, argument parser
and setuptools.

alla3 committed:
Add a Mongo client

Add a Mongo client allowing to manager KNoT Gateway MongoDB database.

alla3 committed:
Add a RabbitMQ client

Add a RabbitMQ client allowing to manager KNoT Gateway RabbitMQ queues.

vitbaq commented on control/__main__.py:
```diff
@@ -0,0 +1,61 @@
+                        default="/usr/local/bin",
+                        type=str,
+                        help="Daemon working directory")
+    parser.add_argument("-p", "--pid-file", metavar="<path/reset>",
```
It's better path/control and tmp/control

ramonhpr replied:
+1 . `--pid-pathname` or `--pid-filepath`

vitbaq commented on control/mongo.py:
```diff
@@ -0,0 +1,48 @@
+        except errors.ServerSelectionTimeoutError as error:
+            logging.error('Err: ' + str(error))
+
+    def drop_all(self):
```
I think that is better drop_all_db. What do you think?

ramonhpr replied:
+1

vitbaq commented on control/rabbitmq.py:
```diff
@@ -0,0 +1,53 @@
+                exceptions.ChannelClosed) as error:
+            logging.error('Err: ' + str(error))
+
+    def remove_all(self):
```
I think that is better remove_all_queue. What do you think?

ramonhpr replied:
+1

vitbaq commented on control/control.py:
```diff
@@ -0,0 +1,47 @@
+        self._rabbitmq_conn = rabbitmq_conn
+
+    def start(self):
+        self._bus_name = dbus.service.BusName('br.org.cesar.knot.control', dbus.SystemBus())
```
Brake line.

vitbaq reviewed (changes requested):
I suggest include some comments on all classes.

ramonhpr commented on setup.py:
```diff
@@ -0,0 +1,35 @@
+    long_description=read("README.md"),
+    entry_points={
+        "console_scripts": [
+            "control = control.__main__:main"
```
what about controld or knot-controld

ramonhpr commented on control/mongo.py:
```diff
@@ -0,0 +1,48 @@
+    def drop_db(self, database):
+        try:
+            logging.info('Removing database: ' + database)
+            self.client.drop_database(database)
```
I think you should verify if `self.client` was assigned. If the function `connect` wasn't called, the control will crash due an `AttributeError`

alla3 replied:
I think it is not desirable to perform a factory reset if it fails.

ramonhpr commented on control/mongo.py:
```diff
@@ -0,0 +1,48 @@
+            self.drop_db(db)
+
+    def close(self):
+        self.client.close()
```
same here

ramonhpr commented on control/rabbitmq.py:
```diff
@@ -0,0 +1,53 @@
+    def remove_queue(self, queue_name):
+        try:
+            logging.info('Removing queue: ' + queue_name)
+            self.ch.queue_purge(queue=queue_name)
```
The same what I sad in `MongoConn` applies here

ramonhpr commented on control/control_factory.py:
```diff
@@ -0,0 +1,19 @@
+    @staticmethod
+    def create(settings):
+        mongodb_settings = settings.get('mongodb')
+        mongo_conm = MongoConn(mongodb_settings.get('host'),
```
`mongo_conn`

ramonhpr commented on control/control_factory.py:
```diff
@@ -0,0 +1,19 @@
+                               mongodb_settings.get('port'))
+        rabbitmq_conn = RabbitMQConn()
+
+        return Control(mongo_conm, rabbitmq_conn)
```
`mongo_conn`

ramonhpr reviewed (changes requested):
You need to create a pattern of the commit. I don't think is good all of them start with `control:` as the repo name is `gateway-control`. You can use like `mongo: rabbitmq:` or `settings: `. The specific component that you are changing. Don't you agree?

alla3 committed:
Add Control Service

Add Control Service allowing to manager KNoT Gateway state. It creates
a D-Bus service and establish connection with others services used by
KNoT Gateway.

jcorde commented on control/settings_factory.py:
```diff
@@ -0,0 +1,20 @@
+
+class SettingsFactory():
+    """
+    Factory to create settings with Reset parameters
```
?

jcorde reviewed (changes requested):
Please, merge the files reset.py and reset_factory.py

jcorde commented on control/control.py:
```diff
@@ -0,0 +1,35 @@
+    """
+
+    def __init__(self):
+        self.reset = ResetFactory.create('settings.json')
```
Please, do not pass this parameter. Use it inside the class as a constant. Also, rename it to something more specific.

jcorde commented on control/control.py:
```diff
@@ -0,0 +1,35 @@
+        self._bus_name = dbus.service.BusName(
+            'br.org.cesar.knot.control', dbus.SystemBus())
+        dbus.service.Object.__init__(
+            self, self._bus_name, '/br/org/cesar/knot/control')
```
Please, use a constant to `'/br/org/cesar/knot/control'` something as `DBUS path name`...

jcorde commented on control/knot-control.conf:
```diff
@@ -0,0 +1,21 @@
+	<deny send_destination="br.org.cesar.knot.control"/>
+	</policy>
+
+	</busconfig>
```
bad indent

jcorde commented on control/knot-control.conf:
```diff
@@ -0,0 +1,21 @@
+        <limit name="max_match_rules_per_connection">2048</limit>
+
+	<policy at_console="true">
+	<allow send_destination="br.org.cesar.knot.control"/>
```
indent

jcorde commented on control/knot-control.conf:
```diff
@@ -0,0 +1,21 @@
+	</policy>
+
+	<policy context="default">
+	<deny send_destination="br.org.cesar.knot.control"/>
```
indent

alla3 committed:
Add Reset module

Add Reset module allowing to perform a factory reset.

ramonhpr commented on control/reset.py:
```diff
@@ -0,0 +1,63 @@
+
+def createReset():
+    """
+        A factory to create a instance of Reset
```
to create an instance

ramonhpr commented on control/control.py:
```diff
@@ -0,0 +1,38 @@
+        self._bus_name = dbus.service.BusName(
+            'br.org.cesar.knot.control', dbus.SystemBus())
+        dbus.service.Object.__init__(
+            self, self._bus_name, DBUS_PATH_NAME)
```
`self.DBUS_PATH_NAME`
or add the constant out of the class

ramonhpr commented on control/control.py:
```diff
@@ -0,0 +1,38 @@
+import os
+
+from reset import createReset
+from reset import Reset
```
You are importing this and it is not using. Besides you won't need this class as the function of createReset is to import create an instance to you.

ramonhpr commented on control/reset.py:
```diff
@@ -0,0 +1,63 @@
+
+class Reset():
+    """
+    Control KNoT Gateway state using a dbus implementation
```
that is not what the class is doing. It is implementing the features the dbus services will call.

ramonhpr commented on control/reset.py:
```diff
@@ -0,0 +1,63 @@
+
+RESET_SETTINGS_FILE = 'resetSettings.json'
+
+def createReset():
```
the file path shouldn't be passed as argument?

alla3 replied:
@jcorde suggested to use it on a defined constant.

jcorde replied:
Why, @ramonhpr ?

ramonhpr replied:
The createReset is acting like an abstract factory (a design pattern to create an instance of a class by its parameter) and depending on the settings we can have another instance. the default one is to connect with a the local database and rabbitmq of the gateway.
We can have always the same path though but most of daemons let the user pass the configuration file or same flag to change the configuration.

jcorde replied:
_connect with a **the** local_

jcorde replied:
I agree, @ramonhpr . It makes total sense to add it as an argument in cases where the value can be changed.
But I see no way or even reason for us to allow it to be modified by now.

As well mentioned by @joaoaneto : both approaches are correct, being it just a matter of preferences.

Even though, I respect your position and believe that if you say this is the best option, it should be.

ramonhpr replied:
If @joaoaneto says that both are correct so I agree with him. And we can let it without modification for now.

ramonhpr commented on control/resetSettings.json:
```diff
@@ -0,0 +1,6 @@
+{
```
I'd call it defaultSettings since it should not just be the settings for the reset but for all the service kcontrold

alla3 replied:
It is been used just for the reset, since the control does not need any parameters on its creation.
